### ISSUE TITLE
FSDP engine: forward_step stores model_output undetached, pinning autograd graphs across micro-batches (multi-GB leak)

### ISSUE THREAD

**Issue by xuefei-wang** [verl-project/verl#7166]
## Summary

`FSDPEngineWithLMHead.forward_step` returns `model_output` **undetached** in the dict that
`forward_backward_batch` accumulates into `output_lst`. For any policy loss that does not backward
through *every* tensor stored in `model_output`, the untraversed entries keep their autograd graphs
(and the tensors saved for backward) alive for the whole mini-batch, since the list is only released
after the last micro-batch.

Memory therefore scales with **mini-batch size**, not micro-batch size — so the usual mitigations
(shrinking `ppo_micro_batch_size_per_gpu`, dynamic token caps, optimizer offload) do not help, and
small smoke tests pass while real runs OOM.

https://github.com/volcengine/verl/blob/main/verl/workers/engine/fsdp/transformer_impl.py#L1525

```python
            output = {
                "model_output": model_output,   # <-- graphs retained
                "loss": loss.detach().item(),
                "metrics": metrics,
            }
            return loss, output
```

## Impact is amplified when the fused CE kernel is unavailable

`logprobs_from_logits` silently selects `logprobs_from_logits_v2` when `flash_attn`'s fused
cross-entropy cannot be imported (`FLAH_ATTN_CROSS_ENTROPY_LOSS_AVAILABLE = False`, set in a bare
`except ImportError` with no warning). Its bf16 branch builds a **per-row `log_softmax`**:

```python
        for row_logits, row_labels in zip(logits, labels, strict=True):
            row_logprobs = F.log_softmax(row_logits, dim=-1)
```

Each of those rows is retained for backward, so combined with the issue above the leak is
(tokens x vocab) sized per micro-batch. This bites hardest on wide-vocabulary models.

Note the fallback is easy to hit unintentionally: a `flash_attn` wheel built without the `ops`
subpackage imports fine for `flash_attn.bert_padding` (used elsewhere in verl) but fails for
`flash_attn.ops.triton.cross_entropy`, silently disabling the memory-efficient path with no log line.

## Evidence

Observed on 4x A6000 48GB with a 248k-vocab model and a distillation-style policy loss that consumes
only top-k log-probs (so backward never traverses the stored `log_probs`). Every configuration OOMed
in `update_actor` at the first or second optimizer step.

`torch.cuda.memory._record_memory_history()` + `_snapshot()` taken at the OOM, grouped by allocating
stack, attributed **32.5 GiB across ~35k live blocks** to:

```
torch_functional.py:196:logprobs_from_logits_v2
  <- torch_functional.py:99:logprobs_from_logits
  <- transformer_impl.py:1297:prepare_model_outputs
  <- transformer_impl.py:1483:forward_step
```

i.e. one ~1 MB fp32 vocab row per token, still alive for every micro-batch of the mini-batch.

## Suggested fixes

1. Detach the stored copy after the loss has been built (the loss keeps its own graph references, so
   this is numerically inert and gradients are unaffected):

```python
            model_output = {
                k: (v.detach() if isinstance(v, torch.Tensor) else v)
                for k, v in model_output.items()
            }
```

   Applying only this took the same workload from OOM to a stable peak with ~18 GB headroom, and
   three full training steps at ~78k tokens/GPU/step completed unchanged numerically.

2. Emit a one-time warning when `FLAH_ATTN_CROSS_ENTROPY_LOSS_AVAILABLE` is False, so users know the
   memory-efficient CE path is disabled rather than discovering it through an OOM.

Happy to open a PR for either if the approach looks right.

## Environment

- verl vendored at `v1.0.0`; both code paths verified **unchanged on `main`** at time of filing
- torch 2.8.0+cu128, FSDP2, bf16, 4x RTX A6000 48GB (sm_86)

**Comment by wuxibin89** [comment]
Fixed in https://github.com/verl-project/verl/pull/6699
